javascript/ql/lib/semmle/javascript/endpoints/EndpointNaming.qll

```diff
@@ -0,0 +1,475 @@
+private predicate isPrivateLike(API::Node node) { isPrivateAssignment(node.asSink()) }
+
+private API::Node getASuccessor(API::Node node, string name, int badness) {
+  isExported(node) and
+  exists(string member |
+    result = node.getMember(member) and
+    not isPrivateLike(node) and
+    if member = "default"
+    then
+      if defaultExportCanBeInterpretedAsNamespaceExport(node)
+      then (
+        badness = 5 and name = ""
+      ) else (
+        badness = 10 and name = "default"
+      )
+    else (
+      name = member and badness = 0
+    )
+  )
+}
+
+private API::Node getAPredecessor(API::Node node, string name, int badness) {
+  node = getASuccessor(result, name, badness)
+}
+
+/**
+ * Gets the predecessor of `node` to use when constructing a qualified name for it,
+ * and binds `name` and `badness` corresponding to the label on that edge.
+ */
+private API::Node getPreferredPredecessor(API::Node node, string name, int badness) {
+  // For root nodes, we prefer not having a predecessor, as we use the package name.
+  not isPackageExport(node) and
+  // Rank predecessors by name-badness, export-distance, and name.
+  // Since min() can only return a single value, we need a separate min() call per column.
+  badness = min(int b | exists(getAPredecessor(node, _, b)) | b) and
+  result =
+    min(API::Node pred, string name1 |
+      pred = getAPredecessor(node, name1, badness)
+    |
+      pred order by distanceFromPackageExport(pred), name1
+    ) and
+  name = min(string n | result = getAPredecessor(node, n, badness) | n)
+}
+
+/**
+ * Holds if values escpin
+ */
+private predicate sinkHasNameCandidate(API::Node sink, string package, string name, int badness) {
+  sink = API::moduleExport(package) and
+  name = "" and
+  badness = 0
+  or
+  exists(API::Node baseNode, string baseName, int baseBadness, string step, int stepBadness |
+    sinkHasNameCandidate(baseNode, package, baseName, baseBadness) and
+    baseNode = getPreferredPredecessor(sink, step, stepBadness) and
+    badness = (baseBadness + stepBadness).minimum(20) and
+    name = join(baseName, step)
+  )
+}
+
+/**
+ * Holds if `(package, name)` is the primary name to associate with `node`.
+ *
+ * `badness` is bound to the associated badness of the name.
+ */
+private predicate sinkHasPrimaryName(API::Node sink, string package, string name, int badness) {
+  badness = min(int b | sinkHasNameCandidate(sink, _, _, b) | b) and
+  package = min(string p | sinkHasNameCandidate(sink, p, _, badness) | p) and
+  name = min(string n | sinkHasNameCandidate(sink, package, n, badness) | n order by n.length(), n)
+}
+
+/**
+ * Holds if `(package, name)` is the primary name to associate with `node`.
+ */
+private predicate sinkHasPrimaryName(API::Node sink, string package, string name) {
+  sinkHasPrimaryName(sink, package, name, _)
+}
+
+/**
+ * Holds if `(package, name)` is an alias for `node`.
+ *
+ * This means it is a valid name for it, but was not chosen as the primary name.
```

Comment:
maybe also mention that the "search" only goes 1 step back. 
I.e. removing 1 path-element gets the primary name for a node.

Reply:
I'll need to refactor the aliasing logic a bit...

Reply:
I've pushed a change that refactors the aliasing logic a bit, and also describes the 1-step aliasing rule.